libbpf-tools/doublefree.c

```diff
@@ -0,0 +1,473 @@
+			   const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG && !env.verbose)
+		return 0;
+
+	return vfprintf(stderr, format, args);
+}
+
+static void sig_int(int signo)
+{
+        exiting = 1;
+}
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	switch (key) {
+	case 'h':
+		argp_state_help(state, stderr, ARGP_HELP_STD_HELP);
+		break;
+	case 'v':
+		env.verbose = true;
+		break;
+	case 'p':
+		errno = 0;
+		env.pid = strtol(arg, NULL, 10);
+		if (errno || env.pid <= 0) {
+			p_err("Invalid PID: %s", arg);
+			argp_usage(state);
+		}
+		break;
+	case 'c':
+		env.command = strdup(arg);
+		ON_MEM_FAILURE(env.command);
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+static int attach_uprobes(struct doublefree_bpf *obj)
+{
+	int err = 0;
+	char libc_path[PATH_MAX] = {};
+
+	err = get_pid_lib_path(1, "c", libc_path, PATH_MAX);
+	if (err) {
+		p_err("Failed to find libc.so, %d", err);
+		return err;
+	}
+	ATTACH_URETPROBE_CHECKED(obj, libc_path, malloc, env.pid);
+	ATTACH_UPROBE_CHECKED(obj, libc_path, free, env.pid);
+	ATTACH_URETPROBE_CHECKED(obj, libc_path, calloc, env.pid);
+	ATTACH_URETPROBE_CHECKED(obj, libc_path, realloc, env.pid);
+	ATTACH_UPROBES_CHECKED(obj, libc_path, posix_memalign, env.pid);
+	ATTACH_URETPROBE_CHECKED(obj, libc_path, memalign, env.pid);
+
+	/* Accept attach failure for following functions */
+	ATTACH_URETPROBE(obj, libc_path, aligned_alloc, env.pid);
+	ATTACH_URETPROBE(obj, libc_path, valloc, env.pid);
+	ATTACH_URETPROBE(obj, libc_path, pvalloc, env.pid);
+	ATTACH_URETPROBE(obj, libc_path, reallocarray, env.pid);
+
+	return 0;
+}
+
+static void print_backtrace(const struct syms* syms, int stackid) {
+	const struct sym *sym = NULL;
+	size_t i = 0;
+	int err = 0;
+	unsigned long *ip = NULL;
+	int sfd = bpf_map__fd(obj->maps.stack_traces);
+
+	ip = calloc(env.perf_max_stack_depth, sizeof(*ip));
+	if (!ip) {
+		p_err("Failed to alloc ip");
+		return;
+	}
+	err = bpf_map_lookup_elem(sfd, &stackid, ip);
+	if (err < 0) {
+		p_err("Failed to lookup elem on sfd");
+		free(ip);
+		return;
+	}
+	for (i = 0; i < env.perf_max_stack_depth && ip[i]; ++i) {
+		printf("\t#%zu %#016lx", i + 1, ip[i]);
+		sym = syms__map_addr(syms, ip[i]);
+		if (sym)
+			printf(" %s+%#lx", sym->name, sym->offset);
+		printf("\n");
+	}
+	printf("\n");
+	free(ip);
+}
+
+static int get_stackid(int fd, unsigned long key) {
+	struct doublefree_info_t val = {};
+	int err = 0;
+
+	err = bpf_map_lookup_elem(fd, &key, &val);
+	if (err < 0) {
+		p_err("Failed to lookup elem on fd");
+		return err;
+	}
+	return val.stackid;
+}
+
+static void handle_event(void *ctx, int cpu, void *data, __u32 data_sz)
+{
+	const struct syms *syms = NULL;
+	struct event *e = data;
+	int stackid = 0;
+	int allocs_fd = bpf_map__fd(obj->maps.allocs);
+	int deallocs_fd = bpf_map__fd(obj->maps.deallocs);
+
+	if (e->err == -1) {
+		p_err("This message should not be printed..");
```

Comment:
How about removing code that shouldn't happen?

Reply:
It's a good point. However, thinking that something will never be executed is just the developer's assumption. In software development, it is not uncommon for code that should not be executed to actually be executed. Especially when contributing code to open source projects, where multiple developers are involved, there is a higher possibility of unintentionally executing code that should not be executed. Therefore, I believe that this code should be kept as a precaution for future mistakes by someone else. However, the message itself seems lengthy and lacks specific details. I will consider possible improvements.

Reply:
I didn't remove the message but changed it.